share/xcode/ext/lcms2-2.1/utils/linkicc/linkicc.1
.\"Shiju P. Nair September 30, 2004
.TH ICCLINK 1 "September 30, 2004"
.SH NAME
icclink - little cms device link generator.
.SH SYNOPSIS
.B icclink
.RI [ options ] " <profiles>"
.SH DESCRIPTION
lcms is a standalone CMM engine, which deals with the color management.
It implements a fast transformation between ICC profiles.
.B icclink 
is little cms device link generator.
.P
Links two or more profiles into a single devicelink profile.
Colorspaces must be paired except Lab/XYZ, that can be interchanged.
.SH OPTIONS
.TP
.B \-8
Creates 8-bit devicelink.
.TP
.B \-b 
Black point compensation.
.TP
.B \-c <0,1,2,3>
Precission (0=LowRes, 1=Normal, 2=Hi-res). [defaults to 1]
.TP
.BI \-d\  description
Description text (quotes can be used).
.TP
.B \-h <0,1,2,3>
Show summary of options and examples.
.TP
.BI \-i\  profile
Input profile (defaults to sRGB).
.TP
.B -k <0..400> 
Ink-limiting in % (CMYK only)
.TP
.BI \-o\  profile
Output devicelink profile. [defaults to 'devicelink.icm']
.TP
.B \-t <0,1,2,3>
Intent (0=Perceptual, 1=Colorimetric, 2=Saturation, 3=Absolute).
.TP
.B \-x 
Creatively, guess deviceclass of resulting profile.
.TP
.nf
Built-in profiles:
	*Lab  -- D50-based CIEL*a*b (PCS)
	*XYZ  -- CIE XYZ (PCS)
	*sRGB -- sRGB color space
	*Gray22- Monochrome of Gamma 2.2
	*Lin2222- CMYK linearization of gamma 2.2 on each channel
.fi					
.SH EXAMPLES
.nf
To create 'devicelink.icm' from a.icc to b.icc:
	icclink a.icc b.icc

To create 'out.icc' from sRGB to cmyk.icc:
	icclink -o out.icc *sRGB cmyk.icc

To create a sRGB input profile working in Lab:
	icclink -x -o sRGBLab.icc *sRGB *Lab

To create a XYZ -> sRGB output profile:
	icclink -x -o sRGBLab.icc *XYZ *sRGB

To create a abstract profile doing softproof for cmyk.icc:
	icclink -t1 -x -o softproof.icc *Lab cmyk.icc cmyk.icc *Lab

To create a 'grayer' sRGB input profile:
	icclink -x -o grayer.icc *sRGB gray.icc gray.icc *Lab

To embed ink limiting into a cmyk output profile:
	icclink -x -o cmyklimited.icc -k 250 cmyk.icc *Lab
.fi
.SH NOTES
For suggestions, comments, bug reports etc. send mail to
info@littlecms.com.
.SH SEE ALSO
.BR jpegicc (1),
.BR tifficc (1),
.BR icc2ps (1),
.BR icctrans (1),
.BR wtpt (1)
.SH AUTHOR
This manual page was written by Shiju p. Nair <shiju.p@gmail.com>,
for the Debian project.


ChangeLog
**Version 1.0.7 (April 17 2012):**
    * IRIDAS .look support
    * ociolutimage utility added (handles image <-> 3dlut)
    * CMake build allows optional reliance on system libraries
    * CMake layout changes for python and nuke installs
    * Bumped internals to yaml 0.3.0, pystring 1.1.2
    * Optimized internal handling of Matrix / Exponent Ops
    * Added INTERP_BEST interpolation option
    * Python config.clearLooks() added
    * Python docs revamp
    * Nuke config-dependent knob values now baked into .nk scripts
    * Nuke OCIOLookTransform gets reload button
    * Nuke nodes get updated help text

**Version 1.0.6 (March 12 2012):**
    * JNI (Java) updates
    * ocioconvert arbitrary attr support

**Version 1.0.5 (Feb 22 2012):**
    * Internal optimization framework added
    * SetLoggingLevel(..) bugfix
    * Python API Documentation / website updates
    * Clang compilation fix

**Version 1.0.4 (Jan 25 2012):**
    * ocio2icc deprecated (functionality merged into ociobakelut)
    * rv integration (beta)
    * nuke: updated channel handling
    * Documentation / website updates

**Version 1.0.3 (Dec 21 2011):**
    * Tetrahedral 3dlut interpolation (CPU only)
    * ociocheck and Config.sanityCheck() have improved validation
    * Mari: updated examples
    * Makefile: misc updates to match distro library conventions

**Version 1.0.2 (Nov 30 2011):**
    * 3D lut processing (cpu) is resiliant to nans
    * Nuke OCIOFileTransform gains Reload buttons
    * Installation on multi-lib *nix systems improved
    * Installation handling of soversion for C++/python improved
    * ociobakelut improvements
    * Initial version of Java bindings (alpha)

**Version 1.0.1 (Oct 31 2011):**
    * Luts with incorrect extension are properly loaded
    * ocio2icc / ociobakelut get --lut option (no longer requires ocio config)
    * DisplayTransform looks do not apply to 'data' passes.

**Version 1.0.0 (Oct 3 2011):**
    * ABI Lockoff for 1.0 branch
    * General API Cleanup (removed deprecated / unnecessary functions)
    * New features can be added, but the ABI will only be extended in a binary compatible manner.
      Profiles written from 1.0 will always be readable in future versions.
    * Fixed Truelight Reading Bug
    * ocio2icc no longer requires ocio config (can provide raw lut(s)


-------------------------------------------------------------------------------


**Version 0.8.7 (Oct 3 2011):**
    * Fixed Truelight Reading Bug

**Version 0.8.6 (Sept 7 2011):**
    * Updated .ocio config reading / writing to be forwards compatibile with 1.0
      (Profiles written in 0.8.6+ will be 1.0 compatible. Compatibility from prior versions is likely, though not guaranteed.)
    * Better logging
    * Added ColorSpace.equalitygroup (makes ColorSpace equality explicit)
    * Substantial Nuke node updates
    * Added support for Iridas .itx read/write
    * Windows Build Support

**Version 0.8.5 (Aug 2 2011):**
    * Nuke OCIODisplay fixed (bug from 0.8.4)
    * Updated Houdini HDL Reader / Writer

**Version 0.8.4 (July 25 2011):**
    * Native Look Support
    * Core / Nuke OCIODisplay supports alpha viewing
    * Added Houdini (.lut) writing
    * Added Pandora (.mga,.m3d) reading
    * Additional internal bug fixes

**Version 0.8.3 (June 27 2011):**
    * OCIO::Config symlink resolution fixed (bugfix)
    * Nuke OCIODisplay knobs use string storage (bugfix)
    * Makefile cleanup
    * Documentation cleanup

**Version 0.8.2 (June 7 2011):**
    * Numerous Windows compatibility fixes
    * Python binding improvements
    * OCIO::Baker / ociobakelut improvements
    * Lut1D/3D do not crash on nans (bugfix)
    * Nuke UI doesnt crash in known corner case (bugfix)

**Version 0.8.1 (May 9 2011):**
    * New roles: TEXTURE_PAINT + MATTE_PAINT
    * Mari API Example (src/mari)
    * FileFormat registry updated to allow Windows + Debug support
    * boost_ptr build compatibility option

**Version 0.8.0 (Apr 19 2011):**
    * ABI Lockoff for stable 0.8 branch
      New features can be added, but the ABI will only be extended in a binary compatible manner
    * Otherwise identical to 0.7.9

-------------------------------------------------------------------------------

**Version 0.7.9 (Apr 18 2011):**
    * Added support for .vf luts
    * Misc. Nuke enhancements
    * Adds optional boost ptr support (backwards compatibility)
    * Makefile enhancements (Nuke / CMAKE_INSTALL_EXEC_PREFIX)
    * cdlTransform.setXML crash fix

**Version 0.7.8 (March 31 2011):**
    * Iridas lut (.cube) bugfix, DOMAIN_MIN / DOMAIN_MAX now obeyed
    * Exposed GPU functions in python (needed for Mari)
    * Nuke OCIODisplay cleanup: Fixed knob names and added envvar support
    * ociobaker cleanup
    * tinyxml ABI visibility cleaned up
    * Removed Boost dependency, tr1::shared_ptr now used instead

**Version 0.7.7 (March 1 2011):**
    * Lut baking API + standalone app
    * Truelight runtime support (optional)
    * Cinespace (3d) lut writing
    * CSP prelut support
    * Boost argparse dependency removed
    * SanityCheck is more exhaustive
    * FileTransform supports relative path dirs
    * Python enhancements (transform kwarg support)
    * Makefile enhancements (OIIO Path)
    * Processor API update (code compatible, binary incompatible)

**Version 0.7.6 (Feb 1 2011):**
    * Updated Config Display API (.ocio config format updated)
    * Added ocio2icc app (ICC Profile Generation)
    * Revamp of ASC CDL, added .cc and .ccc support
    * Documentation Improvements
    * Makefile enhancements (Boost_INCLUDE_DIR, etc)

**Version 0.7.5 (Jan 13 2011):**
    * ociodisplay enhancements
    * gpu display bugfix (glsl profile 1.0 only)
    * Makefile enhancements
    * Nuke installation cleanup
    * Doc generation using sphinx (html + pdf)

**Version 0.7.4 (Jan 4 2011):**
    * Added 'Context', allowing for 'per-shot' Transforms
    * Misc API Cleanup: removed old functions + fixed const-ness

share/xcode/ext/lcms2-2.1/utils/linkicc/linkicc.c
//---------------------------------------------------------------------------------
//
//  Little Color Management System
//  Copyright (c) 1998-2010 Marti Maria Saguer
//
// Permission is hereby granted, free of charge, to any person obtaining 
// a copy of this software and associated documentation files (the "Software"), 
// to deal in the Software without restriction, including without limitation 
// the rights to use, copy, modify, merge, publish, distribute, sublicense, 
// and/or sell copies of the Software, and to permit persons to whom the Software 
// is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in 
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO 
// THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE 
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION 
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION 
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
//---------------------------------------------------------------------------------

#include "utils.h"

// ---------------------------------------------------------------------------------

static char* Description = "Devicelink profile";
static char* Copyright   = "No copyright, use freely";
static int   Intent = INTENT_PERCEPTUAL;
static char* cOutProf    = "devicelink.icc";
static int   PrecalcMode  = 1;
static int   NumOfGridPoints = 0;

static cmsBool BlackPointCompensation = FALSE;

static cmsFloat64Number InkLimit   = 400;
static cmsBool lUse8bits           = FALSE;
static cmsBool TagResult           = FALSE;
static cmsBool KeepLinearization   = FALSE;
static cmsFloat64Number Version    = 4.2;


// The manual
static
int Help(int level)
{
     switch (level) {

     default:
     case 0:

         fprintf(stderr, "\nlinkicc: Links profiles into a single devicelink.\n");     

         fprintf(stderr, "\n");     
         fprintf(stderr, "usage: linkicc [flags] <profiles>\n\n");
         fprintf(stderr, "flags:\n\n");         
         fprintf(stderr, "%co<profile> - Output devicelink profile. [defaults to 'devicelink.icc']\n", SW);     

         PrintRenderingIntents();

         fprintf(stderr, "%cc<0,1,2> - Precision (0=LowRes, 1=Normal, 2=Hi-res) [defaults to 1]\n", SW);     
         fprintf(stderr, "%cn<gridpoints> - Alternate way to set precision, number of CLUT points\n", SW);     
         fprintf(stderr, "%cd<description> - description text (quotes can be used)\n", SW);     
         fprintf(stderr, "%cy<copyright> - copyright notice (quotes can be used)\n", SW);    
         
         fprintf(stderr, "\n%ck<0..400> - Ink-limiting in %% (CMYK only)\n", SW);
         fprintf(stderr, "%c8 - Creates 8-bit devicelink\n", SW);
         fprintf(stderr, "%cx - Creatively, guess deviceclass of resulting profile.\n", SW);
         fprintf(stderr, "%cb - Black point compensation\n", SW);
         fprintf(stderr, "%cl - Use linearization curves (may affect accuracy)\n", SW);
         fprintf(stderr, "%cr<v.r> - Profile version. (CAUTION: may change the profile implementation)\n", SW);
         fprintf(stderr, "\n");    
         fprintf(stderr, "Colorspaces must be paired except Lab/XYZ, that can be interchanged.\n\n");

         fprintf(stderr, "%ch<0,1,2,3> - More help\n", SW);
         break;

     case 1:
         PrintBuiltins();
         break;

     case 2:

         fprintf(stderr, "\nExamples:\n\n"
             "To create 'devicelink.icm' from a.icc to b.icc:\n"
             "\tlinkicc a.icc b.icc\n\n"
             "To create 'out.icc' from sRGB to cmyk.icc:\n"
             "\tlinkicc -o out.icc *sRGB cmyk.icc\n\n"
             "To create a sRGB input profile working in Lab:\n"
             "\tlinkicc -x -o sRGBLab.icc *sRGB *Lab\n\n"
             "To create a XYZ -> sRGB output profile:\n"
             "\tlinkicc -x -o sRGBLab.icc *XYZ *sRGB\n\n"
             "To create a abstract profile doing softproof for cmyk.icc:\n"
             "\tlinkicc -t1 -x -o softproof.icc *Lab cmyk.icc cmyk.icc *Lab\n\n"
             "To create a 'grayer' sRGB input profile:\n"
             "\tlinkicc -x -o grayer.icc *sRGB gray.icc gray.icc *Lab\n\n"
             "To embed ink limiting into a cmyk output profile:\n"
             "\tlinkicc -x -o cmyklimited.icc -k 250 cmyk.icc *Lab\n\n");                     
         break;                       

     case 3:

         fprintf(stderr, "This program is intended to be a demo of the little cms\n"
             "engine. Both lcms and this program are freeware. You can\n"
             "obtain both in source code at http://www.littlecms.com\n"
             "For suggestions, comments, bug reports etc. send mail to\n"
             "info@littlecms.com\n\n");
    }

   exit(0);
}

// The toggles stuff
static
void HandleSwitches(int argc, char *argv[])
{
    int s;

    while ((s = xgetopt(argc,argv,"V:v:xXH:h:8k:K:BbO:o:T:t:D:d:C:c:n:N:f:F:lLy:Y:r:R:")) != EOF) {

    switch (s) {

       case 'V':
       case 'v':
           Verbose = atoi(xoptarg);
           if (Verbose < 0 || Verbose > 3) {
               FatalError("Unknown verbosity level '%d'", Verbose);
           }
           break;

       case '8':
           lUse8bits = TRUE;
           break;

       case 'd':
       case 'D':
           // Doing that is correct and safe: Description points to memory allocated in the command line.
           // same for Copyright and output devicelink.
           Description = xoptarg;
           break;

       case 'y':
       case 'Y':
           Copyright = xoptarg;
           break;


       case 'o':
       case 'O':
           cOutProf = xoptarg;
           break;


       case 't':
       case 'T':
           Intent = atoi(xoptarg);  // Will be validated latter on
           break;

       case 'c':
       case 'C':
           PrecalcMode = atoi(xoptarg);
           if (PrecalcMode < 0 || PrecalcMode > 2) {
               FatalError("Unknown precalc mode '%d'", PrecalcMode);
           }
           break;

       case 'n':
       case 'N':
           if (PrecalcMode != 1) {
               FatalError("Precalc mode already specified");
           }
           NumOfGridPoints = atoi(xoptarg);
           break;

       case 'b':
       case 'B':
           BlackPointCompensation = TRUE;
           break;


       case 'k':
       case 'K':
           InkLimit = atof(xoptarg);
           if (InkLimit < 0.0 || InkLimit > 400.0) {
               FatalError("Ink limit must be 0%%..400%%");
           }
           break;

       case 'r':
       case 'R':
          Version = atof(xoptarg);
          if (Version < 2.0 || Version > 4.3) {
              fprintf(stderr, "WARNING: lcms was not aware of this version, tag types may be wrong!\n");
          }
          break;


       case 'x':
       case 'X': TagResult = TRUE;
           break;

       case 'h':
       case 'H':
           Help(atoi(xoptarg));
           return;

           
       case 'l':
       case 'L': KeepLinearization = TRUE;
           break;

       default:

           FatalError("Unknown option - run without args to see valid ones.\n");          
        }       
    }
}

// Set the copyright and description
static
cmsBool SetTextTags(cmsHPROFILE hProfile)
{
    cmsMLU *DescriptionMLU, *CopyrightMLU;
    cmsBool  rc = FALSE;
    cmsContext ContextID = cmsGetProfileContextID(hProfile);

    DescriptionMLU  = cmsMLUalloc(ContextID, 1);
    CopyrightMLU    = cmsMLUalloc(ContextID, 1);

    if (DescriptionMLU == NULL || CopyrightMLU == NULL) goto Error;

    if (!cmsMLUsetASCII(DescriptionMLU,  "en", "US", Description)) goto Error;
    if (!cmsMLUsetASCII(CopyrightMLU,    "en", "US", Copyright)) goto Error;

    if (!cmsWriteTag(hProfile, cmsSigProfileDescriptionTag,  DescriptionMLU)) goto Error;
    if (!cmsWriteTag(hProfile, cmsSigCopyrightTag,           CopyrightMLU)) goto Error;     

    rc = TRUE;

Error:

    if (DescriptionMLU)
        cmsMLUfree(DescriptionMLU);
    if (CopyrightMLU)
        cmsMLUfree(CopyrightMLU);
    return rc;
}



int main(int argc, char *argv[])
{
    int i, nargs, rc;
    cmsHPROFILE Profiles[257];
    cmsHPROFILE hProfile;
    cmsUInt32Number dwFlags;
    cmsHTRANSFORM hTransform = NULL;

    // Here we are
    fprintf(stderr, "little cms ICC device link generator - v2.0 [LittleCMS %2.2f]\n", LCMS_VERSION / 1000.0);
    fflush(stderr);

    // Initialize
    InitUtils("linkicc");
    rc = 0;
    
    // Get the options
    HandleSwitches(argc, argv);

    // How many profiles to link?
    nargs = (argc - xoptind);
    if (nargs < 1)
        return Help(0); 

    if (nargs > 255) {
        FatalError("Holy profile! what are you trying to do with so many profiles!?");
        goto Cleanup;
    }

    // Open all profiles
    memset(Profiles, 0, sizeof(Profiles));
    for (i=0; i < nargs; i++) {

        Profiles[i] = OpenStockProfile(0, argv[i + xoptind]);
        if (Profiles[i] == NULL) goto Cleanup;      

        if (Verbose >= 1) {
            PrintProfileInformation(Profiles[i]);
        }
    }

    // Ink limiting
    if (InkLimit != 400.0) {        
        cmsColorSpaceSignature EndingColorSpace = cmsGetColorSpace(Profiles[nargs-1]);
        Profiles[nargs++] = cmsCreateInkLimitingDeviceLink(EndingColorSpace, InkLimit);
    }

    // Set the flags
    dwFlags = cmsFLAGS_KEEP_SEQUENCE;
    switch (PrecalcMode) {

        case 0: dwFlags |= cmsFLAGS_LOWRESPRECALC; break;
        case 2: dwFlags |= cmsFLAGS_HIGHRESPRECALC; break;
        case 1: 
            if (NumOfGridPoints > 0)
                dwFlags |= cmsFLAGS_GRIDPOINTS(NumOfGridPoints);
            break;

        default: 
            {
                FatalError("Unknown precalculation mode '%d'", PrecalcMode);
                goto Cleanup;
            }
    }

    if (BlackPointCompensation)
        dwFlags |= cmsFLAGS_BLACKPOINTCOMPENSATION;

    if (TagResult)
        dwFlags |= cmsFLAGS_GUESSDEVICECLASS;

    if (KeepLinearization)
        dwFlags |= cmsFLAGS_CLUT_PRE_LINEARIZATION|cmsFLAGS_CLUT_POST_LINEARIZATION;

    if (lUse8bits) dwFlags |= cmsFLAGS_8BITS_DEVICELINK;

    // Create the color transform. Specify 0 for the format is safe as the transform 
    // is intended to be used only for the devicelink.
    hTransform = cmsCreateMultiprofileTransform(Profiles, nargs, 0, 0, Intent, dwFlags);
    if (hTransform == NULL) {
        FatalError("Transform creation failed");
        goto Cleanup;
    }

    hProfile =  cmsTransform2DeviceLink(hTransform, Version, dwFlags);
    if (hProfile == NULL) {
        FatalError("Devicelink creation failed");
        goto Cleanup;
    }

    SetTextTags(hProfile);
    cmsSetHeaderRenderingIntent(hProfile, Intent);

    if (cmsSaveProfileToFile(hProfile, cOutProf)) {

        if (Verbose > 0) 
            fprintf(stderr, "Ok");
    }
    else 
        FatalError("Error saving file!");

    cmsCloseProfile(hProfile);


Cleanup:

    if (hTransform != NULL) cmsDeleteTransform(hTransform);
    for (i=0; i < nargs; i++) {

        if (Profiles[i] != NULL) cmsCloseProfile(Profiles[i]);
    }

    return rc;     
}
